// @akashic/akashic-engine/lib/auxiliary/EmptyVectorImageAsset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmptyVectorImageAsset = void 0;
var trigger_1 = require("@akashic/trigger");
var EmptyVectorImageAsset = /** @class */ (function () {
    function EmptyVectorImageAsset(id, path, width, height, hint) {
        this.type = "vector-image";
        this.width = 0;
        this.height = 0;
        this.onDestroyed = new trigger_1.Trigger();
        this.id = id;
        this.path = path;
        this.originalPath = path;
        this.width = width;
        this.height = height;
        this.hint = hint;
    }
    EmptyVectorImageAsset.prototype.createSurface = function (_width, _height, _sx, _sy, _sWidth, _sHeight) {
        return null;
    };
    EmptyVectorImageAsset.prototype.inUse = function () {
        return false;
    };
    EmptyVectorImageAsset.prototype.destroy = function () {
        if (this.destroyed()) {
            return;
        }
        this.onDestroyed.destroy();
        this.onDestroyed = undefined;
    };
    EmptyVectorImageAsset.prototype.destroyed = function () {
        return !this.onDestroyed;
    };
    EmptyVectorImageAsset.prototype._load = function (loader) {
        loader._onAssetLoad(this);
    };
    EmptyVectorImageAsset.prototype._assetPathFilter = function (path) {
        return path;
    };
    return EmptyVectorImageAsset;
}());
exports.EmptyVectorImageAsset = EmptyVectorImageAsset;
